### ISSUE TITLE
Add whitespace-related properties to FormattingOptions

### ISSUE THREAD

**Issue by rcjsuen** [microsoft/language-server-protocol#375]
```TypeScript
/**
 * Value-object describing what options formatting should use.
 */
interface FormattingOptions {
	/**
	 * Size of a tab in spaces.
	 */
	tabSize: number;

	/**
	 * Prefer spaces over tabs.
	 */
	insertSpaces: boolean;

	/**
	 * Signature for further properties.
	 */
	[key: string]: boolean | number | string;
}
```
Currently, `FormattingOptions` only defines two properties. I would like to propose that we add some additional properties (names can be reworded obviously) such as the following:

```TypeScript
interface FormattingOptions {
    /**
     * Trim trailing whitespaces on a line.
     */
    trimTrailingWhitespace: boolean;

    /**
     * Insert a newline character at the end of the file if one does not exist.
     */
    insertFinalNewline: boolean;

    /**
     * Trim all newlines after the final newline at the end of the file.
     */
    trimFinalNewlines: boolean;
}
```

**Comment by dbaeumer** [comment]
We need to at least make them optional to not break existing servers.

**Comment by rcjsuen** [comment]
You're right. They should be optional.

What do you think of the names, @dbaeumer? I copy and pasted them from VS Code.

```JavaScript
// When enabled, will trim trailing whitespace when saving a file.
"files.trimTrailingWhitespace": false,

// When enabled, insert a final new line at the end of the file when saving it.
"files.insertFinalNewline": false,

// When enabled, will trim all new lines after the final new line at the end of the file when saving it.
"files.trimFinalNewlines": false,
```

**Comment by rcjsuen** [comment]
> I have updated the protocol and will soon update the spec.

@dbaeumer Where did you update the protocol? I don't see it in the [`protocol.ts`](https://github.com/Microsoft/vscode-languageserver-node/blob/053149a76e7ec147347e41a5847a1cc4ccb3d541/protocol/src/protocol.ts
) file in [Microsoft/vscode-languageserver-node](https://github.com/Microsoft/vscode-languageserver-node/). Were you referring to a different file?

**Comment by dbaeumer** [comment]
This looks like it got lost somehow. I can not remember where :-). I am still open to add these. Are you willing to provide a PR?

**Comment by rcjsuen** [comment]
@dbaeumer No big deal, it is what it is. I will try to take a look see over the weekend...

**Comment by rcjsuen** [comment]
@dbaeumer I found it. Looks like it was in the `vscode-languageserver-types` package. See [here](https://github.com/Microsoft/vscode-languageserver-node/commit/fab28ab226152bff352bf41e60ec1adee030c894).

**Comment by rcjsuen** [comment]
@dbaeumer I've created #735. Please take a look when you get a chance, thanks!

**Comment by dbaeumer** [comment]
Merged into dbaeumer/3.15.
